[#169] 페이머니 입출금과 거래 내역 조회 기능 구현

HaejunJang opened the pull request:
## 요약 (Summary)
- [x] 디자인 레퍼런스(팀별 명세 FE 예시 화면)에 맞게 페이머니 기능 개발
- [x] 페이머니 조회 
- [x] 페이머니 + 등록 계좌 조회
- [x] 페이머니 충전
- [x] 페이머니 출금
- [x] 페이머니 거래 내역 조회

## 🔑 변경 사항 (Key Changes)

- 페이머니 충전과 출금은 비관락을 적용하여 구현
- 페이머니 거래내역 조회의 경우 무한스크롤의 모바일 화면을 고려하여 커서기반 페이징처리로 구현 + 각 페이머니 조건(충전, 출금, 결제 등)에 맞게 필터링

## 📝 리뷰 요구사항 (To Reviewers)

- [ ] 각 기능이 동작하는지 확인부탁드립니다
- [ ] 이때 계좌를 등록하고 하셔야 뒤에 기능 확인이 가능합니다
- [ ] 추가적으로 예외처리를 해야할 부분이 있다면 알려주세요
- [ ] 코드적으로 개선해야하는 부분이 있다면 알려주세요

## 확인 방법 
## 1. 페이머니 조회
<img width="986" height="621" alt="image" src="https://github.com/user-attachments/assets/b1aaccd5-224e-4ad8-a165-476720aafdf7" />

- url : http://localhost:9000/AirBnG/wallet/me/balance GET

## 2. 페이머니 + 등록 계좌 조회
<img width="979" height="935" alt="image" src="https://github.com/user-attachments/assets/bea118d8-5a92-423d-87aa-1a7e1af05101" />

- url : http://localhost:9000/AirBnG/wallet/me/overview GET

## 3. 페이머니 충전 (헤더에 멱등키 담기)
<img width="625" height="573" alt="image" src="https://github.com/user-attachments/assets/bbd6d29d-08b8-4349-8db5-64becd3609a8" />

- url : http://localhost:9000/AirBnG/wallet/me/topup POST

- 주의 사항 
- 충전과 출금은 멱등키(IdemPotency-Key)를 헤더에 담아야 합니다
- 랜덤으로 만든 UUID 예시 : d7b6c2d8-8c3e-4f85-bf59-2a9cf17f4321 
- 요청 헤더 예시
<img width="977" height="104" alt="image" src="https://github.com/user-attachments/assets/1f5bc331-be8d-45d9-b1db-cdd8995bcba0" />

- 이미 처리된 결과의 경우
<img width="337" height="114" alt="image" src="https://github.com/user-attachments/assets/a8b2d64b-1a0d-4acc-9f02-9ba5e6d9b636" />

## 4. 페이머니 출금
<img width="635" height="564" alt="image" src="https://github.com/user-attachments/assets/e25419e5-11b0-4324-8f83-fd7bcbb9c676" />

- url : http://localhost:9000/AirBnG/wallet/me/withdraw POST
- 헤더 : 멱등키(IdemPotency-Key) , JWT토큰 담아서 요청

- 이미 처리된 결과의 경우
<img width="367" height="168" alt="image" src="https://github.com/user-attachments/assets/de85f8c6-b169-4545-b340-33e7dae3055d" />

## 5. 페이머니 거래내역 조회(제일 처음 전체 조회시)
<img width="986" height="856" alt="image" src="https://github.com/user-attachments/assets/5b98914c-8bc1-48e3-b3ff-c0fc3671619d" />

- url : http://localhost:9000/AirBnG/wallet/me/history GET
- 헤더 : JWT토큰

## 5.1 페이머니 필터링 충전만 조회시
<img width="978" height="748" alt="image" src="https://github.com/user-attachments/assets/df88c2c9-0425-479d-8f12-45fb78c7a9b0" />
- url : http://localhost:9000/AirBnG/wallet/me/history?type=TOPUP

- 페이징 확인을 위해서 더미데이터 추가
sql : `INSERT INTO wallet_tx (wallet_id, payment_id, wallet_tx_type, wallet_tx_role, wallet_idem_key, amount, created_at, updated_at)
VALUES
(1, NULL, 'TOPUP', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 1200.00, NOW() - INTERVAL 10 DAY, NOW() - INTERVAL 10 DAY),
(1, NULL, 'WITHDRAW', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 800.00, NOW() - INTERVAL 9 DAY, NOW() - INTERVAL 9 DAY),
(1, NULL, 'PAYMENT', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 2500.00, NOW() - INTERVAL 9 DAY, NOW() - INTERVAL 9 DAY),
(1, NULL, 'REFUND', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 2500.00, NOW() - INTERVAL 8 DAY, NOW() - INTERVAL 8 DAY),
(1, NULL, 'TOPUP', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 5000.00, NOW() - INTERVAL 7 DAY, NOW() - INTERVAL 7 DAY),
(1, NULL, 'PAYMENT', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 2200.00, NOW() - INTERVAL 7 DAY, NOW() - INTERVAL 7 DAY),
(1, NULL, 'WITHDRAW', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 1800.00, NOW() - INTERVAL 6 DAY, NOW() - INTERVAL 6 DAY),
(1, NULL, 'TOPUP', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 3000.00, NOW() - INTERVAL 6 DAY, NOW() - INTERVAL 6 DAY),
(1, NULL, 'PAYMENT', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 1300.00, NOW() - INTERVAL 5 DAY, NOW() - INTERVAL 5 DAY),
(1, NULL, 'REFUND', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 1300.00, NOW() - INTERVAL 5 DAY, NOW() - INTERVAL 5 DAY),
(1, NULL, 'TOPUP', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 700.00, NOW() - INTERVAL 4 DAY, NOW() - INTERVAL 4 DAY),
(1, NULL, 'WITHDRAW', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 900.00, NOW() - INTERVAL 4 DAY, NOW() - INTERVAL 4 DAY),
(1, NULL, 'PAYMENT', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 4000.00, NOW() - INTERVAL 3 DAY, NOW() - INTERVAL 3 DAY),
(1, NULL, 'REFUND', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 4000.00, NOW() - INTERVAL 3 DAY, NOW() - INTERVAL 3 DAY),
(1, NULL, 'TOPUP', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 6000.00, NOW() - INTERVAL 2 DAY, NOW() - INTERVAL 2 DAY),
(1, NULL, 'WITHDRAW', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 2000.00, NOW() - INTERVAL 2 DAY, NOW() - INTERVAL 2 DAY),
(1, NULL, 'PAYMENT', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 1500.00, NOW() - INTERVAL 1 DAY, NOW() - INTERVAL 1 DAY),
(1, NULL, 'REFUND', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 1500.00, NOW() - INTERVAL 1 DAY, NOW() - INTERVAL 1 DAY),
(1, NULL, 'TOPUP', 'CREDIT', UNHEX(REPLACE(UUID(), '-', '')), 3500.00, NOW(), NOW()),
(1, NULL, 'WITHDRAW', 'DEBIT', UNHEX(REPLACE(UUID(), '-', '')), 1800.00, NOW(), NOW()); `

## 5.2 페이징 확인하기 
<img width="980" height="938" alt="image" src="https://github.com/user-attachments/assets/9867bd21-68bf-414c-8f5f-865162c560c2" />

NextCusor 값 확인후 재요청

<img width="906" height="906" alt="image" src="https://github.com/user-attachments/assets/4621e71d-5522-4cdd-ae6a-d214ec61ebad" />

HaejunJang committed:
feat : 본인의 보유 포인트 조회

HaejunJang committed:
feat : 본인의 보유 포인트 조회

HaejunJang committed:
Merge remote-tracking branch 'origin/develop' into AIRBNG-169-BE-페이머니-기능-구현

HaejunJang committed:
fix : 지갑 생성시 status값 생성

develop머지후 wallet에 status컬럼이 생성되어 ACTIVE로 명시

HaejunJang committed:
fix : 지갑 생성시 status값 생성

develop머지후 wallet에 status컬럼이 생성되어 ACTIVE로 명시

HaejunJang committed:
fix : Account에 BaseStatus 추가

HaejunJang committed:
fix : Account에 BaseStatus 추가

HaejunJang committed:
fix : 중복 계좌 등록 수정

HaejunJang committed:
fix : 중복 계좌 등록 수정

HaejunJang committed:
feat : 포인트 + 계좌리스트 조회 개발

HaejunJang committed:
feat : 포인트 + 계좌리스트 조회 개발

HaejunJang committed:
feat : 페이머니 충전 개발

HaejunJang committed:
feat : 페이머니 충전 개발

HaejunJang committed:
fix : Account엔티티 status 삭제

HaejunJang committed:
fix : Account엔티티 status 삭제

HaejunJang committed:
feat : 페이머니 출금 개발

HaejunJang committed:
feat : 페이머니 출금 개발

HaejunJang committed:
feat : 페이머니 거래내역 조회 개발

HaejunJang committed:
feat : 페이머니 거래내역 조회 개발

HaejunJang committed:
Merge remote-tracking branch 'origin/develop' into AIRBNG-169-BE-페이머니-기능-구현

# Conflicts:
#	airbng-pay/src/main/java/com/airbng/pay/service/AccountServiceImpl.java

HaejunJang committed:
Merge branch 'AIRBNG-169-BE-페이머니-기능-구현' of https://github.com/shinhanDsActeam/AirBnG into AIRBNG-169-BE-페이머니-기능-구현

HaejunJang committed:
fix : 명칭 통일 idemPotencykey -> idempotencyKey

HaejunJang committed:
fix : 마이페이지 조회시 포인트값 반환

HaejunJang committed:
Merge remote-tracking branch 'origin/develop' into AIRBNG-169-BE-페이머니-기능-구현

Mouon commented:
어머

kwiyoon reviewed (changes requested):
- 페이머니 입출금 `http://localhost:9000/AirBnG/wallet/me/topup` 테스트 중 해당 예외가 발생하였습니다. 검토 부탁드립니다!!

<img width="436" height="400" alt="image" src="https://github.com/user-attachments/assets/a8d6de6e-8c3e-461d-b32a-2bf28cc75a06" />

<img width="324" height="61" alt="image" src="https://github.com/user-attachments/assets/9cbc776e-cb18-4bda-805e-0fedc53c3937" />
<img width="320" height="66" alt="image" src="https://github.com/user-attachments/assets/8f9d4347-67ce-4bcf-b1e6-1464695e5bc4" />
<img width="288" height="60" alt="image" src="https://github.com/user-attachments/assets/53ad0389-c56c-42cd-bdb1-9de7338b247c" />

kwiyoon commented:
> * 페이머니 입출금 `http://localhost:9000/AirBnG/wallet/me/topup` 테스트 중 해당 예외가 발생하였습니다. 검토 부탁드립니다!!
> 
확인해보니 DTO 유효성 검증 추가가 필요해보입니다. 저희 컨플루언스 기술 블로그 - 유효성 검증 탭 참고해서 추가 부탁드립니다~

HaejunJang committed:
fix : 유효성 검증 추가

HaejunJang commented:
> > * 페이머니 입출금 `http://localhost:9000/AirBnG/wallet/me/topup` 테스트 중 해당 예외가 발생하였습니다. 검토 부탁드립니다!!
> 
> 확인해보니 DTO 유효성 검증 추가가 필요해보입니다. 저희 컨플루언스 기술 블로그 - 유효성 검증 탭 참고해서 추가 부탁드립니다~

<img width="473" height="441" alt="image" src="https://github.com/user-attachments/assets/a596b895-fe94-44a0-845b-26c00ca14b12" />

덕분에 유효성 검증의 필요성을 경험했습니다. 컨플루언서의 내용 참고해서 컨트롤러에서 검증후 잘못 요청시 바로 검증이 되게 개선했습니다.

HaejunJang committed:
fix : param import 변경

HaejunJang committed:
fix : 개발기간에는 엑세스 토큰 기한 1일로 수정

HaejunJang committed:
fix : 불필요 import 제거, 최소 입금 금액 하드 코딩 상수 제거

HaejunJang committed:
fix : 출금 dto 타입 변경, 출금 예외처리 순서 변경

- long 타입은 null이 없기에 기본으로 0을 채워서 잘못 요청시 유효성 검증이 제대로 동작하지 않음
Long 으로 변경
- 잔액 부족 예외처리 보다 먼저 게좌가 본인 계좌인지 예외처리부터 진행

kwiyoon reviewed (approved):
비관락, 멱등성 고려하여 페이머니 입출금 구현해주신 부분 좋습니다~!!!
빠른 피드백 반영도 감사합니다~~

페이징 조회도 최신순으로 잘 돼용~! 수고많으셨습니다 🙌
